capsules/src/virtual_i2c.rs

```diff
@@ -5,35 +5,62 @@
 
 use core::cell::Cell;
 use kernel::common::cells::{OptionalCell, TakeCell};
+use kernel::common::dynamic_deferred_call::{
+    DeferredCallHandle, DynamicDeferredCall, DynamicDeferredCallClient,
+};
 use kernel::common::{List, ListLink, ListNode};
 use kernel::hil::i2c::{self, Error, I2CClient, I2CHwMasterClient};
 
 pub struct MuxI2C<'a> {
     i2c: &'a dyn i2c::I2CMaster,
-    devices: List<'a, I2CDevice<'a>>,
+    smbus: Option<&'a dyn i2c::SMBusMaster>,
+    i2c_devices: List<'a, I2CDevice<'a>>,
+    smbus_devices: List<'a, SMBusDevice<'a>>,
     enabled: Cell<usize>,
-    inflight: OptionalCell<&'a I2CDevice<'a>>,
+    i2c_inflight: OptionalCell<&'a I2CDevice<'a>>,
+    smbus_inflight: OptionalCell<&'a SMBusDevice<'a>>,
+    deferred_caller: &'a DynamicDeferredCall,
+    handle: OptionalCell<DeferredCallHandle>,
 }
 
 impl I2CHwMasterClient for MuxI2C<'_> {
     fn command_complete(&self, buffer: &'static mut [u8], error: Error) {
-        self.inflight.take().map(move |device| {
-            device.command_complete(buffer, error);
-        });
+        if self.i2c_inflight.is_some() {
+            self.i2c_inflight.take().map(move |device| {
+                device.command_complete(buffer, error);
+            });
+        } else if self.smbus_inflight.is_some() {
+            self.smbus_inflight.take().map(move |device| {
+                device.command_complete(buffer, error);
+            });
+        }
         self.do_next_op();
     }
 }
 
 impl<'a> MuxI2C<'a> {
-    pub const fn new(i2c: &'a dyn i2c::I2CMaster) -> MuxI2C<'a> {
+    pub const fn new(
+        i2c: &'a dyn i2c::I2CMaster,
+        smbus: Option<&'a dyn i2c::SMBusMaster>,
```

Comment:
Did anyone test that this actually works if you pass in the same `i2c` as both arguments? This seems like the kind of double-reference thing Rust might have problems with.

Reply:
I compile tested it, Rust seems happy